sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.cs

```diff
@@ -104,6 +111,98 @@ public IScaleMonitor GetMonitor()
             return _scaleMonitor.Value;
         }
 
+        internal class EventProcessorStoredEventsManager : IDisposable
+        {
+            private SemaphoreSlim _storedEventsGuard;
+            private int _maxBatchSize;
+            private int _minBatchSize;
+
+            // This is internal for mocking purposes only.
+            internal ConcurrentDictionary<string, List<EventData>> StoredEvents { get; set; }
+
+            public bool HasStoredEvents
+            {
+                get
+                {
+                    try
+                    {
+                        _storedEventsGuard.Wait();
+                        return !StoredEvents.IsEmpty;
+                    }
+                    finally
+                    {
+                        _storedEventsGuard.Release();
+                    }
+                }
+            }
+
+            public EventProcessorStoredEventsManager(int maxBatchSize, int minBatchSize)
+            {
+                StoredEvents = new ConcurrentDictionary<string, List<EventData>>();
+                _storedEventsGuard = new SemaphoreSlim(1, 1);
+                _maxBatchSize = maxBatchSize;
+                _minBatchSize = minBatchSize;
+            }
+
+            public void PartitionClosing()
+            {
+                try
+                {
+                    _storedEventsGuard.Wait();
+                    StoredEvents.Clear();
+                }
+                finally
+                {
+                    _storedEventsGuard.Release();
+                }
+            }
+
+            public EventData[] ProcessWithStoredEvents(EventProcessorHostPartition partitionContext, List<EventData> events = null, bool timerTrigger = false, CancellationToken cancellationToken = default)
+            {
+                Argument.AssertNotNull(partitionContext, nameof(partitionContext));
+                try
+                {
+                    _storedEventsGuard.Wait(cancellationToken);
+                    events ??= new List<EventData>();
+
+                    var partitionId = partitionContext.PartitionId;
+
+                    var numEventsToProcess = events.Count;
+                    var storedEventsForPartition = StoredEvents.GetOrAdd(partitionId, new List<EventData>());
```

Comment:
If we do move to a `ConcurrentQueue<EventData>` in the stored events set, I think we can unify this codepath to just loop and pop elements until the queue is empty or until we read the max size.